quodlibet: fix build; python.senf: init at 1.5.0

cab404 opened the pull request:
Even with all the progress we've made so far, this train moves on, grinding on software left behind by lack of backward compatibility.

A perfectly good thing becomes forgotten, and then reinvented.

Not today though! c:

## Description of changes

Fixed quodlibet build failure (bump to current master, distutils)
senf: init at 1.5.0

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

natsukium commented on pkgs/development/python-modules/senf/default.nix:
```diff
@@ -0,0 +1,38 @@
+buildPythonPackage rec {
+  pname = "senf";
+  version = "1.5.0";
+  format = "setuptools";
```
```suggestion
  pyproject = true;
```
and add `setuptools` to `build-system` to avoid the legacy build.

> pyproject: Whether the pyproject format should be used. As all other formats are deprecated, you are recommended to set this to true. When you do so, pypaBuildHook will be used, and you can add the required build dependencies from build-system.requires to build-system. Note that the pyproject format falls back to using setuptools, so you can use pyproject = true even if the package only has a setup.py. When set to false, you can use the existing [hooks](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#setup-hooks) or provide your own logic to build the package. This can be useful for packages that don't support the pyproject format. When unset, the legacy setuptools hooks are used for backwards compatibility.

https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonpackage-parameters-buildpythonpackage-parameters

natsukium commented on pkgs/development/python-modules/senf/default.nix:
```diff
@@ -0,0 +1,38 @@
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-kFlFEOqYVBM47YWmwUOPuiCqoqSW+I3y0tNlSFZjjNE=";
```
```suggestion
    hash = "sha256-kFlFEOqYVBM47YWmwUOPuiCqoqSW+I3y0tNlSFZjjNE=";
```

natsukium commented on pkgs/development/python-modules/senf/default.nix:
```diff
@@ -0,0 +1,38 @@
+    sha256 = "sha256-kFlFEOqYVBM47YWmwUOPuiCqoqSW+I3y0tNlSFZjjNE=";
+  };
+
+  nativeCheckInputs = with python.pkgs; [
+    pytest
```
```suggestion
  nativeCheckInputs = [
```

natsukium commented on pkgs/development/python-modules/senf/default.nix:
```diff
@@ -0,0 +1,38 @@
+    hypothesis
+  ];
+
+  pytestFlagsArray = [
+    # Both don't work even with HOME specified...
+    "--deselect=tests/test_api.py::test_getuserdir"
+    "--deselect=tests/test_api.py::test_expanduser_user"
+  ];
```
Please use `disabledTests` instead.
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pytestcheckhook-using-pytestcheckhook

natsukium commented on pkgs/development/python-modules/senf/default.nix:
```diff
@@ -0,0 +1,38 @@
+  buildPythonPackage,
+  fetchPypi,
+  lib,
+  python,
```
Please add the Python package directly to build correctly.

cab404 replied:
What exactly do you mean..?

cab404 replied:
I guess you've meant adding the dependency packages so pythonPackages.callPackage args can override them, IDK

natsukium reviewed (dismissed):
Commits should be separated by tasks.
This PR should have three,
`python312Packages.senf: init at 1.5.0`
`quodlibet: format with nixfmt-rfc-style`
`quodlibet: fix build`

cab404 committed:
pythonPackages.senf: init at 1.5.0

Co-authored-by: OTABI Tomoya <tomoya.otabi@gmail.com>

cab404 commented:
> Commits should be separated by tasks. This PR should have three, `python312Packages.senf: init at 1.5.0` `quodlibet: format with nixfmt-rfc-style` `quodlibet: fix build`

senf does support python2.7, so I did add it to python generally intentionally

cab404 committed:
quodlibet: 4.6.0 -> 4.6.0-unstable-2024-08-08

paveloom commented on pkgs/applications/audio/quodlibet/default.nix:
```diff
@@ -1,136 +1,155 @@
+  ];
 
   src = fetchFromGitHub {
     owner = "quodlibet";
     repo = "quodlibet";
-    rev = "refs/tags/release-${version}";
-    hash = "sha256-dkO/CFN7Dk72xhtmcSDcwUciOPMeEjQS2mch+jSfiII=";
+    rev = "3dcf31dfc8db9806d1f73a47fdabc950d35ded1d";
+    hash = "sha256-8qWuxTvMF6ksDkbZ6wRLPCJK1cSqgGMPac/ht6qVpnA=";
```
Just for reference, this PR switches to an unstable version of Quod Libet to make it work on Python 3.12, while https://github.com/NixOS/nixpkgs/pull/326924 patches the stable one (`4.6.0`).

Consider following [this](https://github.com/NixOS/nixpkgs/blob/eec16b48aa9e98c64b4fdd4a564092ea03a9c8a8/pkgs/README.md?plain=1#L352-L354) convention on what the value of the `version` attribute is in such cases.

cab404 replied:
sure

cab404 commented:
I would really like to get some version of the fix to get merged, be it mine or #326924. I guess also that we can remove @pbogdan from maintainer list as they haven't visited GH in two years or so.

cab404 commented:
@SFrijters should I merge yours?

paveloom commented:
Commits https://github.com/NixOS/nixpkgs/pull/326924/commits/38a8afb747cc952abc440d62bce5ac2451ef1204 and https://github.com/NixOS/nixpkgs/pull/326924/commits/211dc5b6687d765ad636a8914cf14a18d472fb59 from https://github.com/NixOS/nixpkgs/pull/326924 still have changes that are not incorporated into this PR.

So, I see two options here:

1) Merging https://github.com/NixOS/nixpkgs/pull/326924, then rebasing this PR on top of those changes.
2) Cherry-picking the first two or three commits (so, excluding the one commit with the patches) from https://github.com/NixOS/nixpkgs/pull/326924, then closing it.

paveloom commented on pkgs/development/python-modules/senf/default.nix:
```diff
@@ -0,0 +1,37 @@
+  lib,
+  hypothesis,
+  pytestCheckHook,
+}:
+buildPythonPackage rec {
```
```suggestion
}:

buildPythonPackage rec {
```

paveloom commented on pkgs/development/python-modules/senf/default.nix:
```diff
@@ -0,0 +1,37 @@
+    homepage = "https://senf.readthedocs.io/en/latest/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ cab404 ];
+  };
+
+}
```
```suggestion
  };
}
```

cab404 replied:
> 

I like the trailing newline in attrsets :(

paveloom replied:
That's totally fine, just a random style suggestion. In no way this is a blocker.

cab404 commented:
> Commits [38a8afb](https://github.com/NixOS/nixpkgs/commit/38a8afb747cc952abc440d62bce5ac2451ef1204) and [211dc5b](https://github.com/NixOS/nixpkgs/commit/211dc5b6687d765ad636a8914cf14a18d472fb59) from #326924 still have changes that are not incorporated into this PR.
> 
> So, I see two options here:
> 
>     1. Merging [quodlibet: clean up, alternative fix for Python 3.12 #326924](https://github.com/NixOS/nixpkgs/pull/326924), then rebasing this PR on top of those changes.
> 
>     2. Cherry-picking the first two or three commits (so, excluding the one commit with the patches) from [quodlibet: clean up, alternative fix for Python 3.12 #326924](https://github.com/NixOS/nixpkgs/pull/326924), then closing it.

First option looks good to me. I'll go ahead and review/merge it.

SFrijters commented:
@cab404 Thanks for merging my PR. I just tried to rebuild my system and ran into my own issue https://github.com/NixOS/nixpkgs/pull/326924#issuecomment-2227293337 again. I've looked into this a bit more and it gets fixed by https://github.com/quodlibet/quodlibet/commit/7b247a5e0c39006c089efd38ca37e80b527baa23 , so if you are going to move ahead and update to an unstable version, great! Otherwise, I will open a PR to add the above commit as a patch to fix my immediate issue.

cab404 commented:
> @cab404 Thanks for merging my PR. 

<3 Thank you for working on it, I am glad that more than one person uses quodlibet in NixOS nowadays)

> I just tried to rebuild my system and ran into my own issue [#326924 (comment)](https://github.com/NixOS/nixpkgs/pull/326924#issuecomment-2227293337) again. I've looked into this a bit more and it gets fixed by [quodlibet/quodlibet@7b247a5](https://github.com/quodlibet/quodlibet/commit/7b247a5e0c39006c089efd38ca37e80b527baa23) (I think its non-deterministic because of an unsorted `os.walk` when trying to find plugins), so if you are going to move ahead and update to an unstable version, great! Otherwise, I will open a PR to add the above commit as a patch to fix my immediate issue.

I'll go ahead and rebase it — I think that supporting a patchset is more cumbersome than supporting a slow moving maintenance/porting master.

cab404 commented:
@SFrijters what is your opinion, though?

SFrijters commented:
Yeah, there are already so many patches we need to pick to have a (stable) build, imo it makes sense to just use their master for now and hope they do a release at some point.

SFrijters commented on pkgs/applications/audio/quodlibet/default.nix:
```diff
@@ -46,7 +45,7 @@
-python3.pkgs.buildPythonApplication rec {
+python3.pkgs.buildPythonApplication {
   pname = "quodlibet${tag}";
   version = "4.6.0";
```
Probably lost the version update to "unstable-<date>" in the rebase?

cab404 replied:
yeah, nice catch!

paveloom reviewed (approved):
Just for reference, seems like the [`senf`](https://pypi.org/project/senf) Python package isn't referenced anymore in the Nix derivation for Quod Libet, compared to your earlier commits. Seems like the source code of Quod Libet has a [copy](https://github.com/quodlibet/quodlibet/tree/3dcf31dfc8db9806d1f73a47fdabc950d35ded1d/quodlibet/packages/senf) of the [source code](https://github.com/quodlibet/senf/tree/b7400b90e576bac79a7d069a6e2f3b40b9ad19b3/senf) of that Python package, so it works just fine.

That's fine to merge, just might be a bit confusing for a random person stumbling upon it later.

SFrijters commented:
Result of `nixpkgs-review pr 333086` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>19 packages built:</summary>
  <ul>
    <li>python311Packages.senf</li>
    <li>python311Packages.senf.dist</li>
    <li>python312Packages.senf</li>
    <li>python312Packages.senf.dist</li>
    <li>quodlibet</li>
    <li>quodlibet-full</li>
    <li>quodlibet-full.dist</li>
    <li>quodlibet-full.doc</li>
    <li>quodlibet-without-gst-plugins</li>
    <li>quodlibet-without-gst-plugins.dist</li>
    <li>quodlibet-without-gst-plugins.doc</li>
    <li>quodlibet-xine</li>
    <li>quodlibet-xine-full</li>
    <li>quodlibet-xine-full.dist</li>
    <li>quodlibet-xine-full.doc</li>
    <li>quodlibet-xine.dist</li>
    <li>quodlibet-xine.doc</li>
    <li>quodlibet.dist</li>
    <li>quodlibet.doc</li>
  </ul>
</details>

SFrijters reviewed (approved):
All seems to work, but I think before the rebase you also had added `distutils` to the `checkInputs` - it seems that is not necessary, but why was it there before?

cab404 commented:
> All seems to work, but I think before the rebase you also had added `distutils` to the `checkInputs` - it seems that is not necessary, but why was it there before?

I've left @paveloom's [patch](https://github.com/NixOS/nixpkgs/blob/5ef3c4fbb94a8d162128e76af7f701c181aeb13f/pkgs/applications/audio/quodlibet/fix-gdist-python-3.12.patch) instead.

At first I've added `distutils` as a dependency due to build failing on distutils missing, and then I've found out that there's that patch.

I've dug a little deeper, and it turns out both [`distutils`](https://github.com/pypa/setuptools/tree/main/setuptools/_distutils) and [`setuptools`](https://github.com/pypa/setuptools/tree/main/setuptools/_distutils) have pretty much the same code copied over under the hood, which is semi-cursed (you might end up with two different distutils implementations in your python quite easily).

So to avoid having to deal with `setuptools` suddenly and somehow being incompatible with external `distutils` I think having @paveloom's patch is preferable to adding potentially different `distutils` manually.

----
Also, turned out that in pyproject packaging I didn't need `senf` either, so I removed it as well ;D
